### PR TITLE
DOC-3788 AWS CDK topic

### PR DESCRIPTION
**Pull request by michaelcretzman**
https://harness.atlassian.net/browse/DOC-3788

# Harness Developer Pull Request
Thanks for helping us make the Developer Hub better. The PR will be looked at
by the maintainers. 

## What Type of PR is This?

- [ x] Issue
- [ ] Feature
- [ ] Maintenance/Chore

If tied to an Issue, list the Issue(s) here:

* https://harness.atlassian.net/browse/DOC-3788

## House Keeping
Some items to keep track of. Screen shots of changes are optional but would help the maintainers review quicker. 

- [ ] Tested Locally
- [ ] *Optional* Screen Shoot.

### CODE REVIEW

**Review by thisrohangupta — APPROVED** [review]

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 1]
```diff
@@ -0,0 +1,937 @@
...
+AWS Cloud Development Kit (AWS CDK), is an open-source software development framework that allows developers to provision AWS infrastructure resources using familiar programming languages, such as Go, Python, Java, C#, etc. CDK simplifies infrastructure as code (IaC) by abstracting away many of the low-level details and providing a higher-level, programmatic approach.
+
+This topic provides steps on using Harness to provision a target AWS environment or resources using AWS CDK.
+
+## Important notes
+
+- You can add AWS CDK provisioning steps to Harness Deploy and Custom stage types.
+- You can perform ad hoc provisioning or provision the target environment for a deployment as part of the deployment stage.
+
+## AWS permissions required
```
[the remark this answers is not in the document]
Comment (tmacari-harness) [reply docs/continuous-delivery/cd-infrastructure/aws-cdk.md:23]: We can mention something like:   You have to endure that AWS CDK cli will be able to authenticate with desired account and permission. You can set access, secret keys and region as environment variables or let CDK cli to to inherit IAM role form EKS cluster the container step will run.
Ref: https://docs.aws.amazon.com/sdkref/latest/guide/access.html https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 2]
```diff
@@ -0,0 +1,937 @@
...
+The provisioning pipeline steps are configured the same way for both use cases. These steps are covered later in this topic.
+
+For details on Harness provisioning, go to [Provisioning overview](/docs/continuous-delivery/cd-infrastructure/provisioning-overview).
+
+
+### Dynamic provisioning steps for different deployment types
+
+:::note
+
+Currently, the dynamic provisioning documented in this topic is behind the feature flag `CD_NG_DYNAMIC_PROVISIONING_ENV_V2`. Contact [Harness Support](mailto:support@harness.io) to enable the feature.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:58]: This is mentioned twice

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 3]
```diff
@@ -0,0 +1,937 @@
...
+1. **Git Clone step:** Clones the CDK app repository into the CD stage's workspace. 
+  
+  By cloning the repository, you gain access to the necessary code, scripts, or configurations, enabling you to perform various actions and ensure a reliable and controlled deployment.
+  
+  The Git Clone step use case for CDK is described later in this document, but for more general information, go to [Git Clone step](/docs/continuous-delivery/x-platform-cd-features/cd-steps/containerized-steps/git-clone-step).
+1. **AWS CDK Bootstrap step (optional):** Runs the `cdk bootstrap` command.
+
+  This step sets up the necessary AWS resources and environment required for deploying CDK applications in a specific AWS region and AWS account. This command is typically run once per AWS account and region to prepare the environment for CDK deployments. 
+
+  If the step group container already has the necessary AWS resources and environment required for deploying CDK applications, you can omit this step.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:192]: Looks more correct like that:  "If the AWS environment is already bootstrapped  and has the necessary AWS resources required for deploying CDK applications, you can omit this step."

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 4]
```diff
@@ -0,0 +1,937 @@
...
+   1. **Git Clone rollback step:** Typically, this step restores the Git branch/tag/commit of repo used in the last successful deployment. You can also choose to clone and branch/tag/commit as part of rollback.
+   2. **AWS CDK Rollback step:** Rolls back the provisioned resources deployed by the failed **CDK Deploy** step to the state of the resources of the last successful deployment.
+
+These steps are described in detail below.
+
+## Docker image registry connector and image for all steps
+
+The CDK steps in the step group are containerized. In the **Container Registry** and **Image** settings in each step, you must provide a Harness connector to a container registry and an image for the pod the step uses.
+
+Harness provides the `aws-cdk-plugin` base image and custom images for different stacks (Java, .NET, Python, Go, etc.) They are located on the Docker Hub registry [aws-cdk-plugin](https://hub.docker.com/r/harness/aws-cdk-plugin/tags). For example, `harness/aws-cdk-plugin:1.0.0` is the base image that contains the CDK CLI and Node.js and `harness/aws-cdk-plugin:1.0.0-java` is the base image for Java.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:215]: "`harness/aws-cdk-plugin:1.0.0-java` is the base image for Java."  >> "`harness/aws-cdk-plugin:1.0.0-java` is a **custom** image for Java created by harness, you can use it or create you own"

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 5]
```diff
@@ -0,0 +1,937 @@
...
+
+The Git Clone step clones the CDK repo containing your CDK app and adds it to a shared space in the container that can be used by all subsequent steps.
+
+:::note
+
+You can omit the Git Clone step if you have added the app files to the shared space on the container using some other means, such as a script in a Shell Script step.
+
+:::
+
+During rollback, the Git Clone rollback step deletes the Git branch/tag/commit that you used in the Git Clone step.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:236]: Not sure If I got it...

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 6]
```diff
@@ -0,0 +1,937 @@
...
+</details>
+
+
+## AWS CDK Bootstrap step
+
+Runs the `cdk bootstrap` command. For details on bootstrapping, go to [Bootstrapping your AWS environment](https://docs.aws.amazon.com/cdk/v2/guide/cli.html#cli-bootstrap) from AWS.
+
+This step sets up the necessary AWS resources and environment required for deploying CDK applications in a specific AWS region and AWS account. This command is typically run once per AWS account and region to prepare the environment for CDK deployments. 
+
+If the step group container already has the necessary AWS resources and environment required for deploying CDK applications, you can omit this step.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:295]: as I mentioned above "If the AWS environment is already bootstrapped and has the necessary AWS resources required for deploying CDK applications, you can omit this step."

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 7]
```diff
@@ -0,0 +1,937 @@
...
+
+Step settings:
+
+- **Container registry:** A Harness Docker registry connector for the registry hosting the image that you want Harness to run commands on, such as Docker Hub.
+- **Image:** The image to use for this step. For example, `harness/aws-cdk-plugin:1.0.0-java`.
+- **App Path:** The path to the CDK app fetched with the Git Clone step. The Git Clone step listed the app repository in its **Repository Name** setting. **App Path** must include the path to the app folder in that directory.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:301]: >> **App Path** must include the path to **the CDK application folde**r in that directory.

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 8]
```diff
@@ -0,0 +1,937 @@
...
+- **AWS CDK Bootstrap Command Options:** You can add any CDK parameters you can see in the `cdk bootstrap --help` command, just like you would in the `cdk` command-line tool. For example, `-c stack1_name=cdkTest1stack1`. For more information, go to [Parameters](https://docs.aws.amazon.com/cdk/v2/guide/parameters.html) from AWS.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:302]: For example, `-c stack1_name=cdkTest1stack1`.    >>   For example, ` --verbose`.

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 9]
```diff
@@ -0,0 +1,937 @@
...
+
+Runs the `cdk diff` command to compare the specified stack and its dependencies with the deployed stacks.
+
+Step settings:
+
+- **Container registry:** A Harness Docker registry connector for the registry hosting the image that you want Harness to run commands on, such as Docker Hub.
+- **Image:** The image to use for this step. For example, `harness/aws-cdk-plugin:1.0`.0-java.
+- **App Path:** The path to the CDK app fetched with the Git Clone step. The Git Clone step listed the app repository in its **Repository Name** setting. App Path must include the path to the app folder in that directory.
+- **AWS CDK Diff Command Options:** You can add any CDK parameters you can see in the `cdk diff --help` command, just like you would in the `cdk` command-line tool. For example, `-c stack1_name=cdkTest1stack1`. For more information, go to [Parameters](https://docs.aws.amazon.com/cdk/v2/guide/parameters.html) from AWS.
+- **Stack Names:** If you are using a multi-stack app, enter the names of each stack here. For example, if your stack names are `cdkTest1Stack1` and `cdkTest1Stack2`, you would select **Add** and enter two stack names, one for each stack.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:317]: If you are using a multi-stack app, enter the names of each stack here.   >>  Those inputs will be passed to CDK cli. If you are using a multi-stack app, enter the names of each stack you want to passed to `cdk diff` command.

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 10]
```diff
@@ -0,0 +1,937 @@
...
+
+  ![picture 2](static/8be0b8f77211f4eabf068c7b6a19bffb0a1ce86a6fe26b7bc0fc4ed3f1a1d8f3.png)  
+  
+  If you omit a stack name, it can cause a step failure. If your app uses only one stack, you do not need to enter its name.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:321]: If you omit a stack name  >>  If you omit a stack name for a multi stack CDK application

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 11]
```diff
@@ -0,0 +1,937 @@
...
+
+For the remaining settings, see [Step settings common to multiple steps](/docs/continuous-delivery/cd-infrastructure/aws-cdk#step-settings-common-to-multiple-steps) below.
+
+
+<details>
+<summary>CDK Diff step log example</summary>
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:327]: not sure if we should add logs examples in docs,,,   for each step we can just mention that we will run  the corresponding cdk command and print the output in execution logs

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 12]
```diff
@@ -0,0 +1,937 @@
...
+
+## AWS CDK Diff step
+
+Runs the `cdk diff` command to compare the specified stack and its dependencies with the deployed stacks.
+
+Step settings:
+
+- **Container registry:** A Harness Docker registry connector for the registry hosting the image that you want Harness to run commands on, such as Docker Hub.
+- **Image:** The image to use for this step. For example, `harness/aws-cdk-plugin:1.0`.0-java.
+- **App Path:** The path to the CDK app fetched with the Git Clone step. The Git Clone step listed the app repository in its **Repository Name** setting. App Path must include the path to the app folder in that directory.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:315]: everywhere for  **App Path:**.  "App Path must include the path to the app folder in that directory."   >> App Path must include the **path to the CDK application folder** in that directory.

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md (hunk 7 above)
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:301]: The path to the CDK application. For example the Git Clone step cloned the app repository into a directory, this directory should be the App path.

Comment (tmacari-harness) [reply docs/continuous-delivery/cd-infrastructure/aws-cdk.md:301]: we can update that for all below steps

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 13]
```diff
@@ -0,0 +1,937 @@
...
+- **AWS CDK Synth Command Options:** You can add any CDK parameters you can see in the `cdk synthesize --help` command, just like you would in the `cdk` command-line tool. For example, `-c stack1_name=cdkTest1stack1`. For more information, go to [Parameters](https://docs.aws.amazon.com/cdk/v2/guide/parameters.html) from AWS.
+- **Stack Names:** If you are using a multi-stack app, enter the names of each stack here. For example, if your stack names are `cdkTest1Stack1` and `cdkTest1Stack2`, you would select **Add** and enter two stack names, one for each stack.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:387]: "enter the names of each stack here."  >>  "enter the names of each stack you want to synthetize here."

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 14]
```diff
@@ -0,0 +1,937 @@
...
+Exporting template file:  hello-cdk/cdk.out/cdkTest1stack1.template.json
+Exporting template file:  hello-cdk/cdk.out/cdkTest1stack2.template.json
+Adding to output: cdkTest1stack1.template.json
+Adding to output: cdkTest1stack2.template.json
+```
+</details>
+
+### Export and reference JSON templates
+
+If the **Export Template** option is selected, the JSON templates for the stacks in the app are exported to the shared folder on the container.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:418]: "If the **Export Template** option is selected, the JSON templates for the stacks in the app are exported to the shared folder on the container." >>
"After this step, synthetized json templates will we available in cdk.out folder. If the **Export Template** option is selected, the JSON templates for the stacks will exported as step output"

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 15]
```diff
@@ -0,0 +1,937 @@
...
+
+You can reference the JSON template after the step has run using an expression in this format:
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:420]: "You can reference the JSON template after the step has run using an expression in this format:"
>>
"You can reference the JSON template from step output after the step has run using an expression in this format:"

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 16]
```diff
@@ -0,0 +1,937 @@
...
+</details>
+
+### Export and reference JSON templates
+
+If the **Export Template** option is selected, the JSON templates for the stacks in the app are exported to the shared folder on the container.
+
+You can reference the JSON template after the step has run using an expression in this format: 
+
+```
+<+pipeline.stages.STAGE_ID.spec.execution.steps.STEP_GROUP_ID.steps.STEP_ID.output.outputVariables.get("STACK_NAME.template.json")>
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:423]: <+pipeline.stages.STAGE_ID.spec.execution.steps.STEP_GROUP_ID.steps.STEP_ID.output.outputVariables.STACK_NAME>

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 17]
```diff
@@ -0,0 +1,937 @@
...
+```
+
+For example:
+
+```
+<+pipeline.stages.test.spec.execution.steps.test.steps.AwsCdkSynth.output.outputVariables.get("cdkTest1stack2.template.json")>
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:429]: <+pipeline.stages.test.spec.execution.steps.test.steps.AwsCdkSynth.output.outputVariables.cdkTest1stack2>

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 18]
```diff
@@ -0,0 +1,937 @@
...
+
+Runs the `cdk deploy` command to deploy the infrastructure defined in your CDK application to your AWS account.
+
+The CDK Deploy step includes a **Provisioner Identifier** setting to track the provisioning it performs. 
+
+The **Provisioner Identifier** is used by the AWS CDK Rollback step to ensure that rollback is performed on the same infrastructure provisioned in the CDK Deploy step.
+
+Harness determines what to roll back using a combination of **Provisioner Identifier** + Harness account id + Harness org id + Harness project id.
+
+If you've made these settings expressions, Harness uses the values it obtains at runtime when it evaluates the expression.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:521]: "Harness determines what to roll back using a combination of **Provisioner Identifier** + Harness account id + Harness org id + Harness project id.

If you've made these settings expressions, Harness uses the values it obtains at runtime when it evaluates the expression."
>>
"Provisioner Identifier must be unique per provisioned infrastructure at project level"

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 19]
```diff
@@ -0,0 +1,937 @@
...
+### Output variable expressions
+
+After pipeline execution, the CDK Deploy step **Output** tab displays several output variables.
+
+![picture 4](static/29275e4d5bc3a244d2c934e36dd611fb13004a7f3887d277923f22d0236f43b3.png)  
+
+
+#### GIT_COMMIT_ID and LATEST_SUCCESSFUL_PROVISIONING_COMMIT_ID
+
+This is the Git commit Id of the CDK app that was deployed.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:626]: "This is the Git commit Id of the CDK app that was deployed."

--->>>

"After every successful cdk deploy Harness will try to obtain the commit SHA from the App path directory.
This commit Id will be saved and exported as step output: GIT_COMMIT_ID.
Also CDK deploy step will output the commit SHA of latest successful cdk deploy form a previous stage execution if there was one: LATEST_SUCCESSFUL_PROVISIONING_COMMIT_ID.
"

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 20]
```diff
@@ -0,0 +1,937 @@
...
+<+pipeline.stages.STAGE_ID.spec.provisioner.steps.STEP_GROUP_ID.steps.STEP_ID.output.outputVariables.LATEST_SUCCESSFUL_PROVISIONING_COMMIT_ID>
+```
+
+For example:
+
+```
+<+pipeline.stages.s1.spec.provisioner.steps.test.steps.AwsCdkDeploy_1.output.outputVariables.LATEST_SUCCESSFUL_PROVISIONING_COMMIT_ID>
+```
+
+#### Stack(s) outputId
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:640]: --->>  Stack(s) output

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 21]
```diff
@@ -0,0 +1,937 @@
...
+You can reference this value using the expression:
+
+```
+<+pipeline.stages.STAGE_ID.spec.provisioner.steps.STEP_GROUP_ID.steps.STEP_ID.cdkOutput.STACK_NAME.outputId>
+```
+
+For example:
+
+```
+<+pipeline.stages.s1.spec.provisioner.steps.test.steps.AwsCdkDeploy_1.cdkOutput.cdkTest3stack2.outputId>
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:666]:
```
<+pipeline.stages.s1.spec.provisioner.steps.test.steps.AwsCdkDeploy_1.cdkOutput.cdkTest3stack2.outputId>
```
--->>
```
<+pipeline.stages.s1.spec.provisioner.steps.test.steps.AwsCdkDeploy_1.cdkOutput.cdkTest3stack2.BucketNameOutput>
```

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 22]
```diff
@@ -0,0 +1,937 @@
...
+  description: 'Name of the S3 bucket',
+});
+
+```
+
+Each CDK app stack output Id is listed in the CDK Deploy step **Output** tab.
+
+You can reference this value using the expression:
+
+```
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:659]:
```
<+pipeline.stages.STAGE_ID.spec.provisioner.steps.STEP_GROUP_ID.steps.STEP_ID.cdkOutput.STACK_NAME.outputId>
```
--->>
```
<+pipeline.stages.STAGE_ID.spec.provisioner.steps.STEP_GROUP_ID.steps.STEP_ID.cdkOutput.STACK_NAME.OUTPUT_ID>
```

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 23]
```diff
@@ -0,0 +1,937 @@
...
+<+pipeline.stages.s1.spec.provisioner.steps.test.steps.AwsCdkDeploy_1.cdkOutput.cdkTest3stack2.outputId>
+```
+
+
+## AWS CDK Destroy step
+
+Runs the `cdk destroy` command to destroy one or more specified stacks.
+
+You can use this step to destroy ephemeral infrastructure provisioned by the AWS CDK Deploy step, or an stack in the CDK app in the container.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:674]: "You can use this step to destroy ephemeral infrastructure provisioned by the AWS CDK Deploy step, or an stack in the CDK app in the container."
--->>>
"
You can use this step to destroy one or a few stacks defined in CDK application."

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 24]
```diff
@@ -0,0 +1,937 @@
...
+cdkTest1stack1 |   1 | 12:45:56 PM | DELETE_COMPLETE      | AWS::CDK::Metadata          | CDKMetadata/Default (CDKMetadata) 
+
+ ✅  cdkTest1stack1: destroyed
+```
+
+</details>
+
+## AWS CDK rollback steps
+
+Typically, the CDK rollback steps roll back the cloned repo of the CDK app to the last successful commit with a Git Clone step and then remove the infrastructure provisioned by the CDK Deploy step using a CDK Rollback step.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:718]: "Typically, the CDK rollback steps roll back the cloned repo of the CDK app to the last successful commit with a Git Clone step and then remove the infrastructure provisioned by the CDK Deploy step using a CDK Rollback step."

--->>>
"CDK rollback step will run cdk deploy using saved inputs and parameters used at last successful cdk deploy from a previous stage execution. The rollback will reference latest successful deploy by **Provisioner identifier**.

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 25]
```diff
@@ -0,0 +1,937 @@
...
+
+The step group contains the Harness connector to a Kubernetes cluster and namespace hosted in your environment. When the pipeline is run, the step group creates a container inside the cluster. Inside the container, a pod is created for each step in the step group using the image you provide in the step. The steps share a common disk space and can reference the same paths.
+
+When you select AWS CDK as the provisioner on the CD stage **Environment** tab, Harness automatically generates a containerized [step group](/docs/continuous-delivery/x-platform-cd-features/cd-steps/step-groups) in **Rollback** containing the steps needed for the AWS CDK.
+
+### Git Clone rollback step
+
+The Git Clone rollback step is simply a Git Clone step used to roll back the Git repo in the container to the branch, tag, or commit SHA that you want to restore in the case of deployment failure.
+
+Typically, the Git Clone rollback step is used to roll back the app source repo in the container to the last successful commit. You can also add Harness steps to manipulate the repo, such as Shell Script step.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:740]: "Typically, the Git Clone rollback step is used to roll back the app source repo in the container to the last successful commit. You can also add Harness steps to manipulate the repo, such as Shell Script step."

--->>>

After each successful deploy harness is storing only CDK deploy step configuration.
In rollback step group customer have to ensure that CDK application is at the revision they want to rollback.  Git clone step can be added with the specific commit SHA that should be used for rollback.
Also CDK Deploy step output can be used to point at last successful deploy commit SHA: `<+pipeline.stages.s1.spec.provisioner.steps.test.steps.AwsCdkDeploy_1.output.outputVariables.LATEST_SUCCESSFUL_PROVISIONING_COMMIT_ID>`

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 26]
```diff
@@ -0,0 +1,937 @@
...
+
+To ensure that the Git Clone step rolls back to the last successful commit, configure the step as follows:
+
+- **Connector:** Select or add a Harness Git connector for the source control provider hosting the CDK app code repository that you want to use.
+- **Repository Name:**  If the connector's **URL Type** is **Repository**, then **Repository Name** is automatically populated based on the repository defined in the connector's configuration. If the connector's **URL Type** is **Account**, then you must specify the name of the code repository that you want to clone into the stage workspace.
+- **Build Type:** Select the branch, tag, or Git commit SHA of the commit you want to use.
+- **Commit SHA:** If you use, **Git Commit SHA**, you can use the `LATEST_SUCCESSFUL_PROVISIONING_COMMIT_ID` expression from the last *successful* CDK Deploy step. For example, `<+pipeline.stages.s2.spec.provisioner.steps.test.steps.AwsCdkDeploy_2.output.outputVariables.LATEST_SUCCESSFUL_PROVISIONING_COMMIT_ID>`. This expression will ensure that the Git Clone step rolls back to the last successful commit.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:749]: "This expression will ensure that the Git Clone step rolls back to the last successful commit."
--->>>
"In this example, this expression will resolve to commit SHA from the 
latest successful execution of AwsCdkDeploy_2 step from a previous stage. So Git Clone step will checkout at that specific commit SHA"

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 27]
```diff
@@ -0,0 +1,937 @@
...
+
+AWS CDK rollback steps in Harness stages must be added in a [containerized step group](/docs/continuous-delivery/x-platform-cd-features/cd-steps/containerized-steps/containerized-step-groups). The steps cannot be selected outside of a containerized step group.
+
+The step group contains the Harness connector to a Kubernetes cluster and namespace hosted in your environment. When the pipeline is run, the step group creates a container inside the cluster. Inside the container, a pod is created for each step in the step group using the image you provide in the step. The steps share a common disk space and can reference the same paths.
+
+When you select AWS CDK as the provisioner on the CD stage **Environment** tab, Harness automatically generates a containerized [step group](/docs/continuous-delivery/x-platform-cd-features/cd-steps/step-groups) in **Rollback** containing the steps needed for the AWS CDK.
+
+### Git Clone rollback step
+
+The Git Clone rollback step is simply a Git Clone step used to roll back the Git repo in the container to the branch, tag, or commit SHA that you want to restore in the case of deployment failure.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:738]: There is no step type "Git Clone rollback" we have only Git Clone.  I think above sentence may bring confusion so users would look for a "Git Clone rollback" step.

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 28]
```diff
@@ -0,0 +1,937 @@
...
+- **Repository Name:**  If the connector's **URL Type** is **Repository**, then **Repository Name** is automatically populated based on the repository defined in the connector's configuration. If the connector's **URL Type** is **Account**, then you must specify the name of the code repository that you want to clone into the stage workspace.
+- **Build Type:** Select the branch, tag, or Git commit SHA of the commit you want to use.
+- **Commit SHA:** If you use, **Git Commit SHA**, you can use the `LATEST_SUCCESSFUL_PROVISIONING_COMMIT_ID` expression from the last *successful* CDK Deploy step. For example, `<+pipeline.stages.s2.spec.provisioner.steps.test.steps.AwsCdkDeploy_2.output.outputVariables.LATEST_SUCCESSFUL_PROVISIONING_COMMIT_ID>`. This expression will ensure that the Git Clone step rolls back to the last successful commit.
+
+  You do not have to use the Git commit used by the last successful CDK Deploy step. You can rollback to any branch, tag, or commit you like.
+
+<details>
+<summary>CDK Git Clone rollback step log example</summary>
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:754]: I don't think we should add logs here

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 29]
```diff
@@ -0,0 +1,937 @@
...
+
+</details>
+
+### AWS CDK Rollback step
+
+The CDK Rollback step rolls back the provisioned resources deployed by the CDK Deploy step to the last successful version. 
+
+Step settings:
+
+- **Provisioner Identifier:** In the CDK Rollback step, use the identical **Provisioner Identifier** as the CDK Deploy step to ensure that it rolls back the resources deployed by the failed CDK Deploy step.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:789]: "**Provisioner Identifier:** In the CDK Rollback step, use the identical **Provisioner Identifier** as the CDK Deploy step to ensure that it rolls back the resources deployed by the failed CDK Deploy step."
  --->>>
  "Provisioner Identifier is used to link CDK Deploy and CDK Rollback steps.
"

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 30]
```diff
@@ -0,0 +1,937 @@
...
+  By using the same **Provisioner Identifier** in both the CDK Deploy and CDK Rollback step, you do not need to provide identical stack information to ensure that the provisioned resources are rolled back. Harness tracks the provisioned resources from each deployment using the **Provisioner Identifier**, and can roll back to the last successful provisioned resources.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:790]: "By using the same **Provisioner Identifier** in both the CDK Deploy and CDK Rollback step, you do not need to provide identical stack information to ensure that the provisioned resources are rolled back. Harness tracks the provisioned resources from each deployment using the **Provisioner Identifier**, and can roll back to the last successful provisioned resources."
--->>>
"By using the same **Provisioner Identifier** in both the CDK Deploy and CDK Rollback step, you ensure that CDK Rollback will use the data from corresponding cdk deploy". After each successful cdk deploy, Harness is storing the details using the **Provisioner Identifier** so those can be used for rollback.

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md [hunk 31]
```diff
@@ -0,0 +1,937 @@
...
+
+Maximum memory that the container can use. You can express memory as a plain integer or as a fixed-point number with the suffixes G or M. You can also use the power-of-two equivalents, Gi or Mi. Do not include spaces when entering a fixed value. The default is 500Mi.
+
+### Limit CPU
+
+The maximum number of cores that the container can use. CPU limits are measured in CPU units. Fractional requests are allowed. For example, you can specify one hundred millicpu as 0.1 or 100m. The default is 400m. For more information, go to Resource units in Kubernetes.
+
+### Environment Variables
+
+You can change the values of environment variables used in the container. You must input a **Key** and **Value** for each variable.
```
Comment (tmacari-harness) [inline docs/continuous-delivery/cd-infrastructure/aws-cdk.md:926]: "You can change the values of environment variables"
--->>
"You can change or add environment variables"

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelcretzman) [reply docs/continuous-delivery/cd-infrastructure/aws-cdk.md:23]: added

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md (hunk 3 above)
Comment (michaelcretzman) [reply docs/continuous-delivery/cd-infrastructure/aws-cdk.md:192]: added

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md (hunk 4 above)
Comment (michaelcretzman) [reply docs/continuous-delivery/cd-infrastructure/aws-cdk.md:215]: added

File: docs/continuous-delivery/cd-infrastructure/aws-cdk.md (hunk 5 above)
Comment (michaelcretzman) [reply docs/continuous-delivery/cd-infrastructure/aws-cdk.md:236]: clarified it.

**Pull request merged**